sdk/eventhub/event-hubs/src/sender.ts

```diff
@@ -71,12 +99,32 @@ export class EventHubProducer {
    * @throws {TypeError} Thrown if a required parameter is missing.
    * @throws {Error} Thrown if the underlying connection or sender has been closed.
    * @throws {Error} Thrown if a partitionKey is provided when the producer was created with a partitionId.
+   * @throws {Error} Thrown if batch was created with partitionKey different than the one provided in the options.
    * Create a new producer using the EventHubClient createProducer method.
    */
-  async send(eventData: EventData | EventData[], options?: SendOptions): Promise<void> {
+  async send(
+    eventData: EventData | EventData[] | EventDataBatch,
+    options?: SendOptions
+  ): Promise<void> {
     this._throwIfSenderOrConnectionClosed();
     throwTypeErrorIfParameterMissing(this._context.connectionId, "eventData", eventData);
-    if (!Array.isArray(eventData)) {
+
+    if (eventData instanceof EventDataBatch) {
+      throwTypeErrorIfParameterMissing(this._context.connectionId, "events", eventData.events);
+      // throw an error if partition key is different than the one provided in the options.
+      if (options && options.partitionKey !== eventData.partitionKey) {
+        const error = new Error(
+          "Partition key is different than the one provided in the send options."
+        );
+        log.error(
+          "[%s] Partition key is different than the one provided in the send options. %O",
+          this._context.connectionId,
+          error
+        );
+        throw error;
+      }
+      eventData = eventData.events;
+    } else if (!Array.isArray(eventData)) {
       eventData = [eventData];
     }
     return this._eventHubSender!.send(eventData, { ...this._senderOptions, ...options });
```

Comment:
When using EventDataBatch, we're already creating the AMQP message each time an event is added when checking the current message size. With that in mind, it might make sense to refactor EventHubSender a bit to take in a raw AMQP message as well. That way, you don't have to re-encode the message again since you've already done it!